tests/bot/rules/__init__.py

```diff
@@ -0,0 +1,76 @@
+import unittest
+from abc import abstractmethod
+from typing import Callable, Dict, Iterable, List, NamedTuple, Tuple
+
+from tests.helpers import MockMessage
+
+
+class DisallowedCase(NamedTuple):
+    """Encapsulation for test cases expected to fail."""
+    recent_messages: List[MockMessage]
+    culprits: Iterable[str]
+    n_violations: int
+
+
+class RuleTest(unittest.TestCase):
+    """
+    Abstract class for antispam rule test cases.
+
+    Tests for specific rules should inherit from `RuleTest` and implement
+    `relevant_messages` and `get_report`. Each instance should also set the
+    `apply` and `config` attributes as necessary.
+
+    The execution of test cases can then be delegated to the `run_allowed`
+    and `run_disallowed` methods.
+    """
+
+    apply: Callable  # The tested rule's apply function
+    config: Dict[str, int]
```

Comment:
These _could_ be enforced

```py
    @property
    @abc.abstractmethod
    def apply(self) -> Callable:
        """Docstring"""
        raise NotImplementedError
```

Reply:
I briefly considered this, but eventually decided not to. Should the child unittest not provide an `apply` attr, the tests will fail with:

```python
AttributeError: 'MyRuleTest' object has no attribute 'apply'
```

I believe that's descriptive enough. I'll mark this as resolved then if that's ok.